src/feature/hs/hs_dos.c

```diff
@@ -156,8 +156,10 @@ hs_dos_can_send_intro2(or_circuit_t *s_intro_circ)
 {
   tor_assert(s_intro_circ);
 
-  /* Always allowed if the defense is disabled. */
```

Comment:
Hmm this commit was confusing to me. I need to look into this more. But perhaps we can make `param_introduce_defense_enabled` private so that it's obvious that it should not be used by public functions like this one?

Reply:
`param_introduce_defense_enabled` needs to be public from a function. I believe one of the last commits makes it public so the intro point sets the service intro circuit with that value by default to know if it enables the defenses or not.

See `hs_dos_get_intro2_enabled_param()`

Reply:
Hmm, I'm still confused by this. Here are some reasons:

- I don't like the name of `hs_dos_get_intro2_rate_param()` and its two friends. The name I don't like because "param" is an ambiguous word in general, but even more so in this particular context given that the code means "consensus param" but the first thing that came to my mind were the new params from the ESTABLISH_INTRO cell. 

- I don't like that `hs_dos_get_intro2_enabled_param()` is public since it's not clear whether that function knows the ground truth, or I need to consult the `s_intro_circ->introduce2_dos_defense_enabled` to know the ground truth.

- It's a bit confusing to have `get_param_intro_dos_enabled()` and also `hs_dos_get_intro2_rate_param()` and neither of these two things gives me the ground truth in some cases (because if the values are specified in the establish_intro cell the ground truth is different).

Here is a suggestion. Please tell me if you like it:

- Dump the `get_param_intro_dos_enabled` etc. functions and do the consensus check inline in `set_consensus_parameters()`. This way we dont have two similarly named functions doing almost the same thing.

- Abstract the following code of `hs_intropoint.c` into its own `hs_dos_setup_default_introduce2_defences()` `hs_dos` function:


  ```
  circ->introduce2_dos_defense_enabled = hs_dos_get_intro2_enabled_param();
  token_bucket_ctr_init(&circ->introduce2_bucket,
                        hs_dos_get_intro2_rate_param(),
                        hs_dos_get_intro2_burst_param(),
                        (uint32_t) approx_time());
  ```

  This way we can then make the `hs_dos_get_intro2_` functions into private for the hs_dos module.

- Finally, rename the `hs_dos_get_intro2_rate_param()` functions into `get_intro2_rate_default_consensus_value()` because now they are private.

I can also try to do this if you want me , and then we can see if we like it.

Reply:
Ok I've pushed several new commits to address this. Not very complicated. I did not follow exactly what you said, made some changes, you'll see. Main thing is that I kept the consensus parameter variables (see start of `hs_dos.c`) but now I only use those in the hs_dos.c code.

Last commit is also a fix that I found while doing this refactoring.